### DISCUSSION TITLE
spheres_from_coords: Add option to define shape of output image

### DISCUSSION THREAD

**Discussion by heinsimon** [PMEAL/porespy#934] [Ideas]
Hi,

It would be nice, if the spheres_from_coords function would allow to define a image shape, where the spheres given in df are added into.
This could be done as additional mode.

Possible application: 
- Add spheres using pseudo_gravity_packing
- Extract the coordinates of the spheres
- Option A:
  - Remove some of the spheres from this list
  - Create new structure using spheres_from_coords
- Option B:
  - Modify the radius of the spheres
  - Create new structure using spheres_from_coords
  - Combine structure from pseudo_gravity_packing with new structure to obtain particle shells

**Comment by jgostick** [comment]
The main idea of `spheres_from_coords` is that the coords are coming from some other package or source. I like the suggestion you made elsewhere that the pseudo-packing functions should return (optionally?) as list of coords that could be used in the `spheres_from_coords` function. If this were the case, would it address your situation?  

Btw, it's probably relatively easy to write a 'spheres_to_coords' function.  You could use a watershed to find each sphere, find the coords from the center of mass of each spheres, and a local thickness to find their radius, then combine all the results into a data frame.

**Reply by heinsimon** [reply]
I think it was not clear what I meant:
- If spheres in the input of the ```spheres_from_coords``` do not reach x, y or z == 0, I would like to have the option to pad the resulting image to a given shape or add the spheres into a given box.


The following example is not as simple as possible, since ... 
- .. there is one bug in ```pseudo_gravity_packing``` preventing it from adding all possible spheres (3 in this case, instead only adds one)  ...
- and ```spheres_from_coords``` has a bug, that it does not allow for adding just one sphere.
- The positions of the spheres in df are identical to the positions of the spheres of packing_bnd.
- The idea is, that one could define the shape of the image, such, that the result of ```spheres_from_coords``` would be identical to the output ```packing_bnd```

```python
import porespy as ps
import numpy as np
import matplotlib.pyplot as plt

im_box_bnd=np.full((100,50,50),True)
im_box_bnd[0:10,:,:]=False
im_box_bnd[:,0:10,:]=False
im_box_bnd[:,-10:,:]=False
im_box_bnd[:,:,0:10]=False
im_box_bnd[:,:,-10:]=False

im_box_nobnd=np.full((83,30,30),True)

# Bug in pseudo_gravity_packing: To get spheres with radius R one needs to input radius R+1
packing_bnd=ps.generators.pseudo_gravity_packing(im_box_bnd,r=15,clearance=0,maxiter=3,seed=42)
packing_nobnd=ps.generators.pseudo_gravity_packing(im_box_nobnd,r=15,clearance=0,maxiter=3,seed=42)

# Bug in pseudo_gravity_packing: There is room for 3 spheres, but only one is added.
# df={}
# df['X']=np.array([23,51,79])
# df['Y']=np.array([24,23,26])
# df['Z']=np.array([25,26,23])
# df['R']=np.array([14,14,14])

# Bug in spheres_from_coords: Adding only one sphere is not possible
df={}
df['X']=np.array([23,51])
df['Y']=np.array([24,23])
df['Z']=np.array([25,26])
df['R']=np.array([14,14])

packing_sfc=ps.generators.spheres_from_coords(df,smooth=True,mode='contained')
packing_sfc[27:,:,:]=0 # remove second particle, because spheres_from_coords does not allow only one particle

print(f'np.sum bnd: {np.sum(packing_bnd)}')
print(f'np.sum no bnd: {np.sum(packing_nobnd)}')
print(f'np.sum sfc: {np.sum(packing_sfc)}')


print(f'Shape bnd: {packing_bnd.shape}')
print(f'Shape no bnd: {packing_nobnd.shape}')
print(f'Shape sfc: {packing_sfc.shape}')

fig,ax=plt.subplots(1,3, figsize=[12, 6])
ax[0].imshow(np.sum(packing_bnd,axis=1))
ax[0].set_title('packing bnd')
ax[1].imshow(np.sum(packing_nobnd,axis=1))
ax[1].set_title('packing no bnd')
ax[2].imshow(np.sum(packing_sfc,axis=1))
ax[2].set_title('packing spheres from coords')

```

**Reply by heinsimon** [reply]
- #935
- #936
